src/compiler/checker.ts

```diff
@@ -3077,64 +3082,81 @@ namespace ts {
                     return undefined;
                 }
 
-                if (type.flags & TypeFlags.Any) {
+                if (type.flags & TypeFlags.Any || (!(context.flags & NodeBuilderFlags.NoTruncation) && context.approximateLength > (context.tracker.maximumApproximateLength || 2000))) {
+                    context.approximateLength += 3;

```

Comment:
Tracking the max length state at each level of the tree walk is a start, but I think a complete solution would also be able to abort at any level of the tree walk too. Specifically, globally abort and then truncate instead of aborting and truncating locally.

Reply:
I don't understand why the code here doesn't do that, actually. Why don't the short-circuits in witeKind et al implement a global truncation?

Reply:
They're global within the context of a single writer, however signature help composed multiple writer calls to build the full display...

Reply:
Oh no :(